Add Cursor type

zeenix opened the pull request:
This is a no_std variant of `std::io::Cursor`.

Fixes #631.

The implementation was generated by an LLM by copying the std impl and adapting it. I (Zeeshan) reviewed the code and adjusted small bits. The tests added are a further assurance that the code is doing what it's supposed to do. Of course, it needs to be still reviewed by at least another person before we can be 100% certain about it.

zeenix committed:
Add Cursor type

This is a no_std variant of `std::io::Cursor`.

Fixes #631.

The implementation was generated by an LLM by copying the std impl and
adapting it. I (Zeeshan) reviewed the code and adjusted small bits. The
tests added are further guarantee that the code is doing what it's
supposed to do.

jannic commented on embedded-io-adapters/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+/// The standard library implements some I/O traits on various types which
+/// are commonly used as a buffer, like `Cursor<Vec<u8>>` and `Cursor<&[u8]>`.
+///
+/// This is the `embedded-io` equivalent of [`std::io::Cursor`].
```
Maybe this should mention that this also took big parts of the implementation from std?
I'm not sure which amount of attribution is legally required by the licenses, but I think it's at least good style to disclose such copying.

zeenix replied:
> it's at least good style to disclose such copying.

* If it's complicated code and/or it's a specific author(s) to attribute, then I could agree but it's not really the case here.
* In any case, I don't think this should be in user-facing documentation as it doesn't affect users in any way. I'm OK with adding a non-doc comment that says so.
* Is this not true for any other parts of this crate? If it is the case, then we should mention it there too, no? :thinking:

jannic replied:
It's probably more a personal preference than a rule, and I have no idea if other parts of this crate have been copied from elsewhere. (And my wording "it's at least good style to ..." is likely too strong, I should have written something like "I'd prefer to").
And I agree that it shouldn't be in the user-facing documentation. A simple comment "Parts of this file were copied from std::io::Cursor" would do.
If no one else comments on it, just take it as my personal opinion and feel free to ignore my comment.

1-rafael-1 replied:
Just an opinion: A reason to comment copied code is if it turns out to have a bug the person troubleshooting can check if the original code has already been fixed.

1-rafael-1 replied:
Just an opinion: A reason to comment copied code is if it turns out to have a bug the person troubleshooting can check if the original code has already been fixed.

zeenix replied:
Sure, I'll add the comment. Thank you for your detailed feedback. 👍 

Any comments on the changes themselves? I'd like this to be merged soon.

zeenix replied:
Sure, I'll ask the comment. Thank you for your detailed feedback. 👍 

Any comments on the changes themselves? I'd like this to be merged soon.

zeenix commented:
@spookyvision may I ask what makes you 👎 this PR? If it's the use of LLM, perhaps you would have felt fine about this if I hadn't followed the guidelines of the Rust embedded community and hadn't disclosed the use of LLM?

0e4ef622 commented on embedded-io-adapters/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+#[derive(Debug, Default, Eq, PartialEq)]
+pub struct Cursor<T> {
+    inner: T,
+    pos: u64,
```
Why is `pos` a u64 and not a usize?

zeenix replied:
Because the std-API this is modelled after and heavily derived from, [use a u64](https://doc.rust-lang.org/stable/src/std/io/cursor.rs.html#77). I have also wondered sometimes about that but I always assumed that if it's in std, there probably is a good reason for it.

0e4ef622 replied:
Hmm, if there's a good reason, I can't find it. The best I can come up with is, `Seek` uses `u64`/`i64`, so `cursor.seek(SeekFrom::Start(u64::MAX))` shouldn't error. Which isn't very convincing. I also found https://github.com/rust-lang/rfcs/issues/894 regarding this. But I'm not a maintainer here, just my 2 cents.

zeenix replied:
@0e4ef622 Thanks. I'd prefer keeping it as close to std as possible since the point is to provide the exact same API. I appreciate you looking into this though. :pray:

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+/// let mut cursor = Cursor::new(&mut buf[..]);
+///
+/// cursor.write_all(b"some bytes").unwrap();
+/// // Internal buffer is not at capacity so writing more bytes fails.
```
I don't think this comment makes sense. Was it meant to be "Internal buffer is now at capacity"?

You also don't describe *how* it fails.
Does it return a Err? Which one? The `std` version will keep calling write until it's all written, so this difference in behaviour should be documented.

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+    }
+}
+
+// Write implementation for Cursor<&mut [u8]>.
```
Why does this comment exist? It's just restating what the code says in about the same number of words.
Also none of the other impl blocks above got a comment so it's pretty weird that you decided to document this one.

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+
+    use super::{cmp, Cursor};
+
+    // Write implementation for Cursor<Vec<u8>>.
```
This comment provides negative value (it's just the function signature).

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+        }
+    }
+
+    // Write implementation for Cursor<&mut Vec<u8>>.
```
This comment provides negative value (it's just the function signature).

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+
+    use super::{cmp, Cursor, CursorError};
+
+    // Write implementation for Cursor<Box<[u8]>>.
```
This comment provides negative value (it's just the function signature).

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+{
+    /// Returns the remaining slice.
+    #[inline]
+    pub fn remaining_slice(&self) -> &[u8] {
```
In std [they removed `remaining_slice` and `is_empty`](https://github.com/rust-lang/rust/pull/109174) and replaced them with [split](https://doc.rust-lang.org/src/std/io/cursor.rs.html#233-237) and [`split_mut`](https://doc.rust-lang.org/std/io/struct.Cursor.html#method.split_mut).
I think switching to those functions would be an improvement (since `remaining_slice` is equal to `Cursor.split().1`) and `is_empty` is equal to `Cursor.split().1.is_empty()`, but unlike these functions it would not be unambiguous what it means for both readers and writers.

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.remaining_slice().is_empty()
+    }
```
As above, this is `Cursor.split().1.is_empty()` on std.

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+            buf[0] = slice[0];
+        } else {
+            buf[..amt].copy_from_slice(&slice[..amt]);
+        }
```
Is it worth having an optimisation for a read of length 1 here?
Have you checked that this produces better code in the general case?

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+
+        if !buf.is_empty() && len == 0 {
+            return Err(CursorError::Full);
+        }
```
On `std` this would return Ok(0), does it make sense to deviate from that behaviour here?

zeenix replied:
That seems very strange to me and a foot gun (I actually have a vague memory of getting bitten by it). OTOH I also want this API to be as close to that of std as possible so I'm a bit torn here. So I'd welcome a 3rd perspective here.

zeenix replied:
Oh and while looking into this, it turns out that `embedded_io::Write::write_all` has one major difference with `std::io::Write::write_all`: The former panics on `write` returning `Ok(0)`. So I think LLM did the right thing here. 😉

9names commented on embedded-io/src/cursor.rs:
```diff
@@ -0,0 +1,643 @@
+
+            if !buf.is_empty() && amt == 0 {
+                return Err(CursorError::Full);
+            }
```
On std this would return `Ok(0)`, does it make sense to deviate from that behaviour here?

zeenix replied:
See [here](https://github.com/rust-embedded/embedded-hal/pull/717/changes/BASE..35adbdfc825c48260c427c3d5cc811bcd9113a5a#r2653420170) for justification of returning `Error` instead here.

zeenix commented:
I believe I've now addressed all the comments (except for the very unhelpful :-1: reactions from @spookyvision and @jamwaffles, which I can't do anything about).

@Dirbaio Any chance you'd be able to review soon?

MabezDev reviewed (dismissed):
In my opinion this is a solid change, thanks for taking the time to refine the impl.

LGTM, I'll leave the final decision to @Dirbaio as eio is mostly his work.

zeenix commented:
@reitermarkus @MabezDev Thanks so much! I rebased the branch on latest master now.

zeenix commented:
> LGTM, I'll leave the final decision to @Dirbaio as eio is mostly his work.

@MabezDev While I think that's fair, I think we also shouldn't block things on @Dirbaio. They have single-handedly created most of this amazing Rust embedded world after all and they are one person in the end. Moreover, I have pinged them many times over the months about this PR, including after your comment on the Matrix channel and here, so it's safe to say that they don't have time for this right now.

So I'd humbly request that this be merged, unless @Dirbaio raises an objection/cocern. It has received multiple reviews and IMO a pretty straight forward PR.

Dirbaio commented:
Should this be on embedded-io or on embedded-io-adapters? If it's in embedded-io we can never ever make breaking changes to it once released. (I believe we've already discussed this in Matrix?)

Dirbaio commented:
There's some divergence compared to std cursor:

- `impl<const N: usize> Write for Cursor<[u8; N]>` is in std but not in e-io.
- `get_mut`, `set_position` are const in std, not in e-io.
- `remaining_slice` doesn't exist in std. Do we want to diverge from std here? IMO `split()` (which *is* in std) already covers all the use cases of `remaining_slice`.

zeenix commented:
> remaining_slice doesn't exist in std. Do we want to diverge from std here? IMO split() (which is in std) already covers all the use cases of remaining_slice.

`remaining_slice` did exist in the first iteration but was removed: https://github.com/rust-embedded/embedded-hal/pull/717#discussion_r2637561642


---

P.S. I'm AFK for a week so I'll address your concerns when I'm back.

Dirbaio commented:
> remaining_slice did exist in the first iteration but was removed

it wasn't removed, it's still there

zeenix commented:
> Should this be on embedded-io or on embedded-io-adapters? If it's in embedded-io we can never ever make breaking changes to it once released. (I believe we've already discussed this in Matrix?)

IMO it belongs here (embedded-io) but I don't foresee any breaking changes in this API. Perhaps that's why it's probably best we discuss this API to death to ensure all is good. :) However, if you or other maintainers think that it's best in the adapters crate, I'm happy to move it.

> > remaining_slice did exist in the first iteration but was removed
> 
> it wasn't removed, it's still there

Ah right. Sorry, I was already on holidays brain. :laughing: Since you're the second person to say this here, I guess we can remove it, yeah.

Dirbaio commented:
> IMO it belongs here (embedded-io) but I don't foresee any breaking changes in this API. Perhaps that's why it's probably best we discuss this API to death to ensure all is good. :) However, if you or other maintainers think that it's best in the adapters crate, I'm happy to move it.

generally we've tried to keep the trait crates as traits only. The adapters could also be in embedded-io but they're in a separate crate due to this.

We can discuss on Tue in the meeting.

zeenix commented:
> * `impl<const N: usize> Write for Cursor<[u8; N]>` is in std but not in e-io.
> * `get_mut`, `set_position` are const in std, not in e-io.
> * `remaining_slice` doesn't exist in std. Do we want to diverge from std here? IMO `split()` (which _is_ in std) already covers all the use cases of `remaining_slice`.

All addressed, I believe.

zeenix commented:
> * `get_mut`, `set_position` are const in std, not in e-io.

Oh, turns out there was a reason for this: [our MSRV](https://github.com/rust-embedded/embedded-hal/actions/runs/24395388055/job/71251777105?pr=717#step:4:112).

zeenix commented:
> > * `get_mut`, `set_position` are const in std, not in e-io.
> 
> Oh, turns out there was a reason for this: [our MSRV](https://github.com/rust-embedded/embedded-hal/actions/runs/24395388055/job/71251777105?pr=717#step:4:112).

I made them non-const again for now. Let me know if you'd rather bump the MSRV and have these const.

I also took the liberty of refactoring a bit to avoid code duplication in all the writing impls.

Dirbaio commented:
i'd just bump the MSRV yes

zeenix committed:
Bump MSRV to 1.83

zeenix commented:
> i'd just bump the MSRV yes

Done. What was the consensus on where to put this in the meeting? It was hard to follow the last meeting. :laughing:

zeenix commented:
@Dirbaio also, you'll need to update the name of the required MSRV check in the settings here.

zeenix committed:
Rename MSRV job

zeenix commented:
@Dirbaio @adamgreig Hi, as agreed, I moved this to -adapters. PTALA.

Also remove the version from the MSRV job so kindly update the settings again. :pray:
